### PR TITLE
new(modern_probe): Add getsockopt

### PR DESCRIPTION
**Pull request by dwindsor**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

/area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

> /area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

#723 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dwindsor** [comment]
Sorry for the large number of useless commits made during this rebasing. Apparently, resolving conflicts using the Github webgui creates a merge commit.

**Comment by Andreagit97** [comment]
Hey @dwindsor I've pushed some stuff on your branch but feel free to discard the changes if you don't like them. The issue was that `optlen` in `getsockopt` is a pointer so we need to manage it in a different way with respect to `setsockopt`. I've also added a test to check this behavior :) 

```c
int getsockopt(int sockfd, int level, int optname, void *restrict optval, socklen_t *restrict optlen);
```

BTW the PR looks good to me, we just need a rebase to resolve conflicts but I will wait for it, let's try to go in order with your PRs and merge them one by one in the order you opened them, otherwise, you will spend your days resolving conflicts :smile:

**Comment by Andreagit97** [comment]
In the meanwhile i will try to review all of them :)

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getsockopt.bpf.c [hunk 1]
```diff
@@ -68,9 +68,15 @@ int BPF_PROG(getsockopt_x,
...
 	auxmap__store_u8_param(auxmap, sockopt_optname_to_scap(level, optname));
 
+	/* `optval` and `optlen` will be the ones provided by the user if the syscall fails
+	 * otherwise they will be directly extracted from the socket.
+	 */
+
 	/* Parameter 5: optval (type: PT_DYN) */
 	unsigned long optval = extract__syscall_argument(regs, 3);
-	u16 optlen = (u16)extract__syscall_argument(regs, 4);
+	u16 optlen = 0;
```
Comment (hbrueckner) [inline driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getsockopt.bpf.c:77]: Based on the syscall definition:
````c
SYSCALL_DEFINE5(getsockopt, int, fd, int, level, int, optname,
                char __user *, optval, int __user *, optlen)
```
`u16` should be an `int`.

So far I could `grep` `socklen_t` and `__socklen_t` are 32-bit wide in size.

Comment (hbrueckner) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getsockopt.bpf.c:77]: Other than that, this also looks good to me.

Comment (Andreagit97) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getsockopt.bpf.c:77]: great catch!

**Comment by dwindsor** [comment]
Ugh, I used the incorrect merge algorithm when pulling.

File: driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getsockopt.bpf.c (hunk 1 above)
Comment (dwindsor) [reply driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/getsockopt.bpf.c]: Fixed, thanks

**Review by Andreagit97 — APPROVED** [review]
Thank you
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**